File: python/gopax.py
# pylint: disable=invalid-name,missing-docstring,line-too-long,multiple-imports
# pylint: disable=too-many-arguments,redefined-outer-name,too-few-public-methods,deprecated-lambda
"""
GOPAX RESTful API Wrapper
"""
import time, base64, hashlib, hmac, json
from requests.auth import AuthBase
import requests


class GopaxApiAuth(AuthBase):
    def __init__(self, api_key, secret_key):
        self.api_key = api_key
        self.secret_key = secret_key

    def __call__(self, request):
        nonce = str(time.time())
        endpoint = request.path_url.split('?')[0]

	print request.body

        message = nonce + request.method + endpoint + (request.body or '')

        hmac_key = base64.b64decode(self.secret_key)
        signature = hmac.new(hmac_key, message, hashlib.sha512)
        signature_b64 = base64.b64encode(signature.digest())

        request.headers.update({'API-Key': self.api_key,
                                'Signature': signature_b64,
                                'Nonce': nonce})
        return request


class GopaxApi(object):
    def __init__(self, host, api_key, secret_key, verbose=False):
        self.host = host
        self.auth = GopaxApiAuth(api_key, secret_key)
        self.verbose = verbose

    def _invoke_gopax_api(self, method, endpoint, params=None, body=None):
        url = 'https://' + self.host + endpoint
        resp = method(url, params=params, json=body, auth=self.auth)

        if self.verbose:
            print '{} {}'.format(resp.request.method, resp.request.path_url)
            print 'Headers: {}'.format(resp.request.headers)
            print 'Body: {}'.format(resp.request.body)

        if not resp.ok:
            print '{} {}: {}'.format(resp.status_code, resp.reason, resp.text)
            return None
        return resp.json()

    # API List (Authentication needed)
    def GetBalances(self, asset=None):
        return self._invoke_gopax_api(requests.get, '/balances/' + (asset or ''))

    def GetOrders(self, trading_pair_name=None):
        params = {}
        if trading_pair_name:
            params['trading-pair-name'] = trading_pair_name
        return self._invoke_gopax_api(requests.get, '/orders', params=params)

    def GetOrder(self, order_id):
        return self._invoke_gopax_api(requests.get, '/orders/{}'.format(order_id))

    def PlaceOrder(self, trading_pair_name, limit_or_market, side, price, amount):
        body = {'tradingPairName': trading_pair_name,
                'type': limit_or_market,
                'side': side,
                'price': price,
                'amount': amount}
        return self._invoke_gopax_api(requests.post, '/orders', body=body)

    def CancelOrder(self, order_id):
        return self._invoke_gopax_api(requests.delete, '/orders/{}'.format(order_id))

    def GetTrades(self, trading_pair_name=None):
        params = {}
        if trading_pair_name:
            params['trading-pair-name'] = trading_pair_name
        return self._invoke_gopax_api(requests.get, '/trades', params=params)

    # API List (Public)
    def GetAssets(self):
        return self._invoke_gopax_api(requests.get, '/assets')

    def GetTradingPairs(self):
        return self._invoke_gopax_api(requests.get, '/trading-pairs')

    def GetTicker(self, trading_pair_name):
        return self._invoke_gopax_api(requests.get, '/trading-pairs/{}/ticker'.format(trading_pair_name))

    def GetOrderbook(self, trading_pair_name, level=None):
        params = {}
        if level:
            params['level'] = level
        return self._invoke_gopax_api(
            requests.get, '/trading-pairs/{}/book'.format(trading_pair_name), params=params)

    def GetRecentTrades(self, trading_pair_name, limit=100, pastmax=None,
                        latestmin=None, after=None, before=None):
        params = {'limit': limit,
                  'pastmax': pastmax,
                  'latestmin': latestmin,
                  'after': after,
                  'before': before}

        params = dict(filter(lambda (k, v): v is not None, params.items()))
        return self._invoke_gopax_api(
            requests.get, '/trading-pairs/{}/trades'.format(trading_pair_name), params=params)

    def GetStats(self, trading_pair_name):
        return self._invoke_gopax_api(requests.get, '/trading-pairs/{}/stats'.format(trading_pair_name))

    def GetHistoricalData(self, trading_pair_name, start, end, interval):
        params = {'start': start, 'end': end, 'interval': interval}
        return self._invoke_gopax_api(
            requests.get, '/trading-pairs/{}/candles'.format(trading_pair_name), params=params)


# Example
if __name__ == '__main__':

    host = 'api.gopax.co.kr'
    api_key = '<your api key>'
    secret_key = '<your secret key>'

    api = GopaxApi(host, api_key, secret_key)
    print api.GetBalances()
    print api.PlaceOrder('ETH-KRW', 'limit', 'buy', 1, 10)

File: php/lib/gopax.php
<?php
namespace Gopax;

class Response
{
    public $httpStatus;
    public $data;
    public $errorMsg;
    public $errorCode;
    
    public function __construct($httpStatus, $json)
    {
        $response = json_decode($json);
        
        $this->httpStatus = $httpStatus;
        
        if ($response->errormsg) {
            $this->errorMsg = $response->errormsg;
            
            $tokenizedError  = explode(':', $response->errormsg);
            $this->errorCode = trim($tokenizedError[1]);
        } else {
            $this->data     = $response;
            $this->errormsg = NULL;
        }
    }
}

class Constants
{
    const LIMIT = 'limit';
    const MARKET = 'market';
    
    const BUY = 'buy';
    const SELL = 'sell';
    
    const BUY_SELL_1 = 1;
    const BUY_SELL_50 = 2;
    
    const PASTMAX = 'pastmax';
    const LATESTMIN = 'latestmin';
    const AFTER = 'after';
    const BEFORE = 'before';
}

class Errors
{
    const INVALID_ASSET = 100;
    const INVALID_TRADING_PAIR = 101;
    const INVALID_USER = 102;
    const INVALID_ORDER_TYPE = 103;
    const NOT_ENABLED_TRADING_PAIR = 104;
    const NOT_ACTIVATED_TRADING_PAIR = 105;
    const NOT_ENABLED_ASSET = 106;
    const INVALID_AMOUNT = 107;
    const INVALID_PRICE = 108;
    const TOO_MANY_ACTIVE_ORDERS = 200;
    const INSUFFICIENT_BALANCE = 201;
    const INVALID_ID = 202;
    const INVALID_NUMBERS_OVERFLOW = 203;
}

class Request
{
    public $data;
    
    public function toJSON()
    {
        return json_encode($this->data);
    }
}

class OrderRequest extends Request
{
    public function __construct(string $type, string $side, float $price, float $amount, string $tradingPairName)
    {
        $this->data['type']            = $type; // LIMIT, MARKET
        $this->data['side']            = $side; // BUY, SELL
        $this->data['price']           = $price;
        $this->data['amount']          = $amount;
        $this->data['tradingPairName'] = $tradingPairName;
    }
}

class Client
{
    private $apiKey;
    private $apiSecret;
    
    const API_HOST = 'https://api.gopax.co.kr';
    const VERSION = 'gopax-php-sdk-20171216';
    
    public function __construct(string $apiKey = '', string $apiSecret = '')
    {
        $this->apiKey    = $apiKey;
        $this->apiSecret = $apiSecret;
    }
    
    private static function getNonce()
    {
        $mt = explode(' ', microtime());
        return $mt[1] . substr($mt[0], 2, 6);
    }
    
    private function getSignature($nonce, $method, $path, $body)
    {
        $tokenizedPath = explode('?', $path);
        $requestPath   = $tokenizedPath[0];
        $data          = $nonce . $method . $requestPath . $body;
        $secret        = base64_decode($this->apiSecret);
        return base64_encode(hash_hmac('sha512', $data, $secret, true));
    }
    
    private function request(string $method, string $path, Request $request = NULL)
    {
        $curl = curl_init();
        
        $nonce     = self::getNonce();
        $method    = strtoupper($method);
        $postData  = $request ? $request->toJSON() : '';
        $signature = $this->getSignature($nonce, $method, $path, $postData);
        
        $headers[] = 'Content-Type: application/json';
        $headers[] = 'API-KEY: ' . $this->apiKey;
        $headers[] = 'SIGNATURE: ' . $signature;
        $headers[] = 'NONCE: ' . $nonce;

        curl_setopt($curl, CURLOPT_USERAGENT, self::VERSION);
        curl_setopt($curl, CURLOPT_URL, self::API_HOST . $path);
        curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($curl, CURLOPT_HTTPHEADER, $headers);
        
        if ($method === 'POST') {
            curl_setopt($curl, CURLOPT_POST, TRUE);
            curl_setopt($curl, CURLOPT_POSTFIELDS, $postData);
        } elseif ($method !== 'GET') {
            curl_setopt($curl, CURLOPT_CUSTOMREQUEST, $method);
        }
        
        $json       = curl_exec($curl);
        $httpStatus = curl_getinfo($curl, CURLINFO_HTTP_CODE);
        curl_close($curl);
        
        return new Response($httpStatus, $json);
    }
    
    public function getBalances(string $assetName = NULL)
    {
        $url = '/balances';
        
        if ($assetName !== NULL) {
            $url .= "/$assetName";
        }
        
        return $this->request('GET', $url);
    }
    
    public function getOrders(string $tradingPairName = NULL)
    {
        $url = '/orders';
        
        if ($tradingPairName !== NULL) {
            $url .= "?trading-pair-name=$tradingPairName";
        }
        
        return $this->request('GET', $url);
    }
    
    public function getOrder(int $orderId)
    {
        return $this->request('GET', "/orders/$orderId");
    }
    
    public function order(OrderRequest $request)
    {
        return $this->request('POST', "/orders", $request);
    }
    
    public function cancelOrder(int $orderId)
    {
        return $this->request('DELETE', "/orders/$orderId");
    }
    
    public function getTrades(string $tradingPairName = NULL)
    {
        $url = '/trades';
        
        if ($tradingPairName !== NULL) {
            $url .= "?trading-pair-name=$tradingPairName";
        }
        
        return $this->request('GET', $url);
    }
    
    public function getAssets()
    {
        return $this->request('GET', '/assets');
    }
    
    public function getTradingPairs()
    {
        return $this->request('GET', '/trading-pairs');
    }
    
    public function getTicker(string $tradingPairName)
    {
        return $this->request('GET', "/trading-pairs/$tradingPairName/ticker");
    }
    
    public function getOrderbook(string $tradingPairName, int $level = NULL)
    {
        $url = "/trading-pairs/$tradingPairName/book";
        
        if ($level !== NULL) {
            $url .= "?level=$level";
        }
        
        return $this->request('GET', $url);
    }
    
    public function getStats(string $tradingPairName)
    {
        return $this->request('GET', "/trading-pairs/$tradingPairName/stats");
    }
    
    public function getRecentTrades(string $tradingPairName, array $options = NULL)
    {
        $url = "/trading-pairs/$tradingPairName/trades";
        
        if ($options) {
            $params = array();
            
            foreach ($options as $key => $value) {
                $params[] = '$key=' . urlencode($value);
            }
            
            if (count($params) > 0) {
                $url .= '?' . join('&', $params);
            }
        }
        
        return $this->request('GET', $url);
    }
    
    public function getCandles(string $tradingPairName, int $start, int $end, int $interval)
    {
        return $this->request('GET', "/trading-pairs/$tradingPairName/candles?" . "start=$start&end=$end&interval=$interval");
    }
}
